z3: init 4.8.15, 4.7.1

ConnorBaker opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

This PR updates the z3 package to the latest release (4.8.15, `z3` is aliased to `z3_4_8`), and adds a release for 4.7.1 (`z3_4_7`)~, 4.6.0 (`z3_4_6`), and 4.5.0 (`z3_4_5`)~.

The `default.nix` file was refactored to reduce redundancy (similar to what was done for the `hadoop` package).

Additionally, `doCheck` was set to `true` and `checkPhase` added to build and run Z3's unit tests.

I set up some GitHub runners for macOS (`x86_64-darwin`) and Ubuntu (`x86_64-linux`) to verify the packages built successfully and passed their unit tests: <https://github.com/ConnorBaker/nixpkgs/actions>.

I've also personally built `z3_4_8` and `z3_4_7`~, `z3_4_6` and `z3_4_5`~ on `aarch64-linux` (an AWS `m6g.metal` instance), `aarch64-darwin` (my M1 MacBook Pro), and `x86_64-linux` (an AWS `m6i.metal` instance).

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Running the command to test compilation of the packages which depend on z3, I got the following on (this was done on an AWS `mg6.metal` instance, which is `aarch64-linux`, but I saw the same on my `aarch64-darwin` machine, and judging from the errors, will see them on `x86_64-linux` as well):

```console
[ec2-user@ip-10-0-0-67 nixpkgs]$ nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"
$ git -c fetch.prune=false fetch --no-tags --force https://github.com/NixOS/nixpkgs master:refs/nixpkgs-review/0
remote: Enumerating objects: 112, done.
remote: Counting objects: 100% (76/76), done.
remote: Total 112 (delta 76), reused 76 (delta 76), pack-reused 36
Receiving objects: 100% (112/112), 109.96 KiB | 3.44 MiB/s, done.
Resolving deltas: 100% (82/82), completed with 38 local objects.
From https://github.com/NixOS/nixpkgs
   2557f81d7db..4b992f5806e  master     -> refs/nixpkgs-review/0
$ git worktree add /home/ec2-user/.cache/nixpkgs-review/rev-0489a21577032f24bad6e90fcaaf445a32e013c4-1/nixpkgs 4b992f5806e29397ca862ec7380faa0d2d7f8c1f
Preparing worktree (detached HEAD 4b992f5806e)
Updating files: 100% (30015/30015), done.
HEAD is now at 4b992f5806e Merge pull request #164876 from TredwellGit/linux
$ nix-env --option system aarch64-linux -f /home/ec2-user/.cache/nixpkgs-review/rev-0489a21577032f24bad6e90fcaaf445a32e013c4-1/nixpkgs -qaP --xml --out-path --show-trace
$ git merge --no-commit --no-ff 0489a21577032f24bad6e90fcaaf445a32e013c4
Auto-merging pkgs/top-level/all-packages.nix
Automatic merge went well; stopped before committing as requested
$ nix-env --option system aarch64-linux -f /home/ec2-user/.cache/nixpkgs-review/rev-0489a21577032f24bad6e90fcaaf445a32e013c4-1/nixpkgs -qaP --xml --out-path --show-trace --meta
3 packages added:
z3_4_5 (init at 4.5.0) z3_4_6 (init at 4.6.0) z3_4_7 (init at 4.7.1)

32 packages updated:
Boogie cppcheck Dafny echidna freeorion fstar hal-hardware-analyzer libbap-master ocaml4.13.1-bap ocamlPackages.z3 (4.8.14 → 4.8.15) petrinizer python3.10-angr python3.10-angrcli python3.10-angrop python3.10-bap python3.10-claripy python3.10-manticore python3.9-angr python3.9-angrcli python3.9-angrop python3.9-bap python3.9-claripy python3.9-manticore rsstail solc sqlitecpp vampire python39Packages.z3 (4.8.14 → 4.8.15) python310Packages.z3 (4.8.14 → 4.8.15) z3 (4.8.14 → 4.8.15) z3-tptp (4.8.14 → 4.8.15) zz-unstable

$ nix --experimental-features nix-command build --no-link --keep-going --option build-use-sandbox relaxed -f /home/ec2-user/.cache/nixpkgs-review/rev-0489a21577032f24bad6e90fcaaf445a32e013c4-1/build.nix
error: builder for '/nix/store/25phgm77hkv1m8gzh9s563m1ga6lm90a-python3.9-angrcli-1.2.0.drv' failed with exit code 1;
       last 10 log lines:
       >             else:
       > >               raise AttributeError(k)
       > E               AttributeError: rax
       >
       > /nix/store/5il8nyf89w120swgsmxzcyj4pkc28c8j-python3.9-angr-9.1.12332/lib/python3.9/site-packages/angr/state_plugins/view.py:79: AttributeError
       > =========================== short test summary info ============================
       > FAILED tests/test_derefs.py::test_max_depth - AttributeError: rax
       > FAILED tests/test_derefs.py::test_loop - AttributeError: rax
       > FAILED tests/test_derefs.py::test_cc - AttributeError: rax
       > ================== 3 failed, 4 passed, 4 deselected in 3.82s ===================
       For full logs, run 'nix log /nix/store/25phgm77hkv1m8gzh9s563m1ga6lm90a-python3.9-angrcli-1.2.0.drv'.
error: builder for '/nix/store/whmv50qfm6n121lc3jvchxcwrzgjxv0b-python3.10-wasm-1.2.drv' failed with exit code 1;
       last 10 log lines:
       >     from .wasmtypes import *
       >   File "/nix/store/r4ycn8dmscrv3b4c1681z2j8hm50p0k8-python3.10-wasm-1.2/lib/python3.10/site-packages/wasm/wasmtypes.py", line 4, in <module>
       >     from .types import UIntNField, UnsignedLeb128Field, SignedLeb128Field
       >   File "/nix/store/r4ycn8dmscrv3b4c1681z2j8hm50p0k8-python3.10-wasm-1.2/lib/python3.10/site-packages/wasm/types.py", line 296, in <module>
       >     class Structure(WasmField):
       >   File "/nix/store/r4ycn8dmscrv3b4c1681z2j8hm50p0k8-python3.10-wasm-1.2/lib/python3.10/site-packages/wasm/compat.py", line 26, in wrapper
       >     return metaclass(cls.__name__, cls.__bases__, orig_vars)
       >   File "/nix/store/r4ycn8dmscrv3b4c1681z2j8hm50p0k8-python3.10-wasm-1.2/lib/python3.10/site-packages/wasm/types.py", line 264, in __new__
       >     isinstance(cur_field, collections.Callable) or
       > AttributeError: module 'collections' has no attribute 'Callable'
       For full logs, run 'nix log /nix/store/whmv50qfm6n121lc3jvchxcwrzgjxv0b-python3.10-wasm-1.2.drv'.
error: 1 dependencies of derivation '/nix/store/x4wlnm98vinljnjhyscr8ddrqwxvwnam-python3.10-manticore-0.3.6.drv' failed to build
error: builder for '/nix/store/a8ji0y14x543hmwxq7c5yxggylichfzq-python3.9-manticore-0.3.6.drv' failed with exit code 1;
       last 10 log lines:
       >   /nix/store/y1jrg5mffhag6v0kca847nq3lq1mz5y4-python3.9-wasm-1.2/lib/python3.9/site-packages/wasm/types.py:264: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.10 it will stop working
       >     isinstance(cur_field, collections.Callable) or
       >
       > -- Docs: https://docs.pytest.org/en/stable/warnings.html
       > =========================== short test summary info ============================
       > FAILED tests/native/test_armv7unicorn.py::Armv7UnicornInstructions::test_ldm
       > FAILED tests/native/test_armv7unicorn.py::Armv7UnicornInstructions::test_ldm_wb
       > FAILED tests/native/test_armv7unicorn.py::Armv7UnicornInstructions::test_pop_multops
       > FAILED tests/native/test_armv7unicorn.py::UnicornConcretization::test_arm_constant
       > = 4 failed, 8561 passed, 63 skipped, 1468 deselected, 2 xfailed, 3 warnings in 195.59s (0:03:15) =
       For full logs, run 'nix log /nix/store/a8ji0y14x543hmwxq7c5yxggylichfzq-python3.9-manticore-0.3.6.drv'.
error: builder for '/nix/store/f4fh7gww0qziw9wkxxnk10i5z2zw0xlj-vampire-4.6.1.drv' failed with exit code 2;
       last 10 log lines:
       >   265 |     return getUniquePersistentIterator(vi(new SubtermIterator(sel.term(), true)));
       >       |                                                               ~~~~~~~~^~
       > Kernel/EqHelper.cpp: In static member function 'static Kernel::TermIterator Kernel::EqHelper::getRewritableSubtermIterator(Kernel::Literal*, const Kernel::Ordering&) [with SubtermIterator = Kernel::NarrowableSubtermIt]':
       > Kernel/EqHelper.cpp:265:71: warning: 'sel' may be used uninitialized in this function [-Wmaybe-uninitialized]
       >   265 |     return getUniquePersistentIterator(vi(new SubtermIterator(sel.term(), true)));
       >       |                                                               ~~~~~~~~^~
       > Kernel/EqHelper.cpp: In static member function 'static Kernel::TermIterator Kernel::EqHelper::getRewritableVarsIterator(Lib::DHSet<unsigned int>*, Kernel::Literal*, const Kernel::Ordering&)':
       > Kernel/EqHelper.cpp:228:84: warning: 'sel' may be used uninitialized in this function [-Wmaybe-uninitialized]
       >   228 |   return getUniquePersistentIterator(vi(new RewritableVarsIt(unstableVars, sel.term(), true)));
       >       |                                                                            ~~~~~~~~^~
       For full logs, run 'nix log /nix/store/f4fh7gww0qziw9wkxxnk10i5z2zw0xlj-vampire-4.6.1.drv'.
error: builder for '/nix/store/vx8b3qmilg411dksv1l9ac8fxcy8nxzh-z3-4.8.15.drv' failed with exit code 1;
       last 10 log lines:
       > unpacking sources
       > unpacking source archive /nix/store/h42hw6j32lmz7nsi05rhqdfh3r8430sk-source
       > source root is source
       > patching sources
       > applying patch /nix/store/ya66s05kf8rxfrnlslkfg9w778jbmbg7-ocamlfind.patch
       > patching file scripts/mk_util.py
       > Reversed (or previously applied) patch detected!  Assume -R? [n]
       > Apply anyway? [n]
       > Skipping patch.
       > 1 out of 1 hunk ignored -- saving rejects to file scripts/mk_util.py.rej
       For full logs, run 'nix log /nix/store/vx8b3qmilg411dksv1l9ac8fxcy8nxzh-z3-4.8.15.drv'.
error: 1 dependencies of derivation '/nix/store/l7m20crm3vlcwnmijzdbwqs8g06kwv03-ocaml4.13.1-z3-4.8.15.drv' failed to build
error: 1 dependencies of derivation '/nix/store/jgiqvff0a6mj7rdx1qfrrwkp3b9ma2vl-ocaml4.13.1-bap-2.2.0.drv' failed to build
error: 1 dependencies of derivation '/nix/store/b2x4ji3xbs62wjm8w6ydf0vb3s08vy6m-libbap-master-2020-11-25.drv' failed to build
error: 1 dependencies of derivation '/nix/store/aq8r85dc33kkvy0dp385pprlw9n1wlh3-python3.10-bap-1.3.1.drv' failed to build
error: 1 dependencies of derivation '/nix/store/isb6ra4394kz4n2s4a8fdwmj0x2vn3q8-python3.9-bap-1.3.1.drv' failed to build
error: builder for '/nix/store/gdrz0y6vqwpirc3f30lv1i3p9x015055-python3.10-angrcli-1.2.0.drv' failed with exit code 1;
       last 10 log lines:
       > ../../nix/store/vnr3mi0vly6cadcw94j4gha2ambix524-python3.10-unicorn-1.0.2-rc4/lib/python3.10/site-packages/unicorn/unicorn.py:5
       >   /nix/store/vnr3mi0vly6cadcw94j4gha2ambix524-python3.10-unicorn-1.0.2-rc4/lib/python3.10/site-packages/unicorn/unicorn.py:5: DeprecationWarning: The distutils.sysconfig module is deprecated, use sysconfig instead
       >     import distutils.sysconfig
       >
       > -- Docs: https://docs.pytest.org/en/stable/warnings.html
       > =========================== short test summary info ============================
       > FAILED tests/test_derefs.py::test_max_depth - AttributeError: rax
       > FAILED tests/test_derefs.py::test_loop - AttributeError: rax
       > FAILED tests/test_derefs.py::test_cc - AttributeError: rax
       > ============ 3 failed, 4 passed, 4 deselected, 2 warnings in 3.77s =============
       For full logs, run 'nix log /nix/store/gdrz0y6vqwpirc3f30lv1i3p9x015055-python3.10-angrcli-1.2.0.drv'.
error: 10 dependencies of derivation '/nix/store/xf0mqkc42abryfg8k65lr5lag7bm82z1-review-shell.drv' failed to build
10 packages failed to build:
libbap ocamlPackages.bap ocamlPackages.z3 python310Packages.angrcli python310Packages.bap python310Packages.manticore python39Packages.angrcli python39Packages.bap python39Packages.manticore vampire

25 packages built:
boogie cppcheck dafny echidna freeorion fstar hal-hardware-analyzer petrinizer python310Packages.angr python310Packages.angrop python310Packages.claripy python310Packages.z3 python39Packages.angr python39Packages.angrop python39Packages.claripy python39Packages.z3 rsstail solc sqlitecpp z3 z3-tptp z3_4_5 z3_4_6 z3_4_7 zz

$ nix-shell /home/ec2-user/.cache/nixpkgs-review/rev-0489a21577032f24bad6e90fcaaf445a32e013c4-1/shell.nix

[nix-shell:~/.cache/nixpkgs-review/rev-0489a21577032f24bad6e90fcaaf445a32e013c4-1]$ 
```  

EDIT: Removed mentions of 4.6.0 and 4.5.0 per my later commit dropping them.

ConnorBaker committed:
z3: init 4.8.15, 4.7.1, 4.6.0, 4.5.0

ConnorBaker commented:
@thoughtpolice and @ttuegel if you have a chance, would you please review and let me know if there are any changes you'd like to see?

vbgl commented:
Can you please also fix `ocamlPackages.z3` as follows:
  - remove the file `pkgs/development/ocaml-modules/z3/ocamlfind.patch`
  - apply the following patch:
~~~
--- a/pkgs/development/ocaml-modules/z3/default.nix
+++ b/pkgs/development/ocaml-modules/z3/default.nix
@@ -7,11 +7,6 @@ else
 let z3-with-ocaml = (z3.override {
   ocamlBindings = true;
   inherit ocaml findlib zarith;
-}).overrideAttrs (o: {
-  patches = (o.patches or []) ++ [
-    # Fix build; see: https://github.com/Z3Prover/z3/issues/5776
-    ./ocamlfind.patch
-  ];
 }); in
 
 stdenv.mkDerivation {
~~~

ConnorBaker committed:
ocamlPackages.z3: Remove unecessary patch

Reported in https://github.com/Z3Prover/z3/issues/5776 and fixed in https://github.com/Z3Prover/z3/commit/4f6fcf8ea78492e1d90398bcd3a663f8517f2a66.

ConnorBaker commented:
> Can you please also fix `ocamlPackages.z3` as follows:
> 
> * remove the file `pkgs/development/ocaml-modules/z3/ocamlfind.patch`
> * apply the following patch:
> 
> ```
> --- a/pkgs/development/ocaml-modules/z3/default.nix
> +++ b/pkgs/development/ocaml-modules/z3/default.nix
> @@ -7,11 +7,6 @@ else
>  let z3-with-ocaml = (z3.override {
>    ocamlBindings = true;
>    inherit ocaml findlib zarith;
> -}).overrideAttrs (o: {
> -  patches = (o.patches or []) ++ [
> -    # Fix build; see: https://github.com/Z3Prover/z3/issues/5776
> -    ./ocamlfind.patch
> -  ];
>  }); in
>  
>  stdenv.mkDerivation {
> ```

Just committed https://github.com/NixOS/nixpkgs/pull/165730/commits/9c83a461e5789eb22c8a6f4c28f632e058cc8410.

ConnorBaker committed:
z3: remove 4.6.0 and 4.5.0

ConnorBaker commented:
I removed 4.6.0 and 4.5.0 from the versions I added, instead opting to keep just the latest version of 4.8 and the last release of 4.7.

Thoughts or suggestions are welcome!

ConnorBaker commented:
Bumping -- @thoughtpolice and @ttuegel any thoughts?

YorikSar commented on pkgs/applications/science/logic/z3/default.nix:
```diff
@@ -1,6 +1,7 @@
 { lib
 , stdenv
 , fetchFromGitHub
+, fetchpatch
```
This is unused as it seems.

ConnorBaker replied:
Yes it is! That was from a prior iteration where I had included 4.6 and 4.5, both of which needed a patch to build. I'll remove it.

ConnorBaker committed:
z3: Remove unused fetchpatch
